[Docs] Database Access for Elastic Guide

alexfornuto opened the pull request:
Fixes #16963 

This PR adds a guide for connecting to an Elasticsearch database through Teleport Database Access. In a separate commit (for easier review) I also sorted the lists of DB access guides alphabetically.

Because Elastic access was adding in 10.3, this guide should not be backported past v10.

Commit:
sort list of db guides alphabetically

Commit:
add Elastic guide

zmb3 commented on docs/config.json:
```diff
@@ -709,6 +709,10 @@
               "slug": "/database-access/guides/mongodb-atlas/"
             },
+            {
+              "title": "Self-Hosted Elastic",
```
Elastic or Elasticsearch? Is there a difference or do we just use them interchangeably?

Tener replied:
AFAIK `Elasticsearch` is the canonical name we should use, Elastic is ambiguous and possibly confusing.

alexfornuto replied:
```suggestion
              "title": "Self-Hosted Elasticsearch",
```

zmb3 commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+
+(!docs/pages/includes/database-access/token.mdx!)
+
+Install Teleport on the host where you will run the Teleport Database Service. Make sure you've selected your installation type (OSS, Enterprise, Clo):
```
```suggestion
Install Teleport on the host where you will run the Teleport Database Service. Make sure you've selected your installation type (OSS, Enterprise, Cloud):
```

zmb3 commented on docs/pages/database-access/guides/elastic.mdx:
````diff
@@ -0,0 +1,248 @@
+<TabItem label="Config File">
+Add the following block to `teleport.yaml`
+```yaml
+db_service:
````
We should probably add the auth server (well, proxy in this case) address here too, as well as the join token.

alexfornuto replied:
Thanks! Is v3 of the config out for Teleport v10?

zmb3 replied:
Teleport 11 only

alexfornuto replied:
So we'll need to remember to adjust this in the backport.

zmb3 commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+
+<Admonition type="note">
+
+The `--auth-server` flag must point to the Teleport cluster's Proxy Service
```
cc @r0mant and @ryanclark: we should eventually add `--proxy` or `--proxy-server` and make `--auth-server` an alias for these. It's really confusing to tell users to set `--auth-server` to the address of the proxy.

alexfornuto replied:
Should we split this comment into a separate issue?

zmb3 commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+
+- `elastic.cas` with Teleport's certificate authority
+- `elastic.key` with a generated private key
+- `elastic.crt` with a generated user certificate
```
Is this a user certificate or a host certificate?

alexfornuto replied:
Updated to say host certificate.

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+
+## Prerequisites
+
+- A self-hosted Elasticsearch database. Elastic cloud [does not support client certificates](https://www.elastic.co/guide/en/cloud/current/ec-restrictions.html#ec-restrictions-security).
```
```suggestion
- A self-hosted Elasticsearch database. Elastic Cloud [does not support client certificates](https://www.elastic.co/guide/en/cloud/current/ec-restrictions.html#ec-restrictions-security), which are required for setting up Database Access.
```
Suggested:
- Using the official capitalization
- Connecting the note about client certs to Database Access more explicitly

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+- A self-hosted Elasticsearch database. Elastic cloud [does not support client certificates](https://www.elastic.co/guide/en/cloud/current/ec-restrictions.html#ec-restrictions-security).
+
+- A host where you will run the Teleport Database Service. Teleport version 10.3
+  or newer must be installed. <ScopedBlock scope={["oss", "enterprise"]}> This can also be the same instance of Teleport running your Auth and/or Proxy service(s).</ScopedBlock>
```
I think we can remove the "Teleport version 10.3
  or newer must be installed" note and the Installation link since we include instructions for installing Teleport on the Database Service host later.

alexfornuto replied:
@ptgott what if a reader already has a Teleport instance running the Database Access service for other databases and wants to add Elasticsearch to it, and is using a version >10.3?

ptgott replied:
I think we can make this more explicit, maybe something like:

```markdown
- A host where you will install and run the Teleport Database Service. 
  If you are already running the Teleport Database Service, you must 
  ensure that it uses Teleport version 10.3 or newer in order to connect to
  Elasticsearch. 
```

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+  --token=/tmp/token \
+  --auth-server=teleport.example.com:3080 \
+  --name=myelastic \
+  --protocol=redis \
```
Just double-checking this is correct

alexfornuto replied:
```suggestion
  --protocol=elastic \
```

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+
+</Admonition>
+
+<Admonition type="tip">
```
Do we need this tip if we have the the "Config File" tab above?

alexfornuto replied:
This tip was in the Redis guide that I used as a template, and then I added the Tab above. If we like the Tab I'll remove this, or vice versa.

I'll address this after we resolve @zmb3's two comments above.

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+
+(!docs/pages/includes/database-access/create-user.mdx!)
+
+Alternately, you can edit an existing user with database access.
```
So a user only needs to scroll through one set of instructions, what would you think of making the user creation/modification instructions separate `TabItem`s in a `Tabs` component?

alexfornuto replied:
I agree that would be better. My concern was modifying the partial file being included here without knowing everywhere else it's being used and how that would affect other docs.

alexfornuto replied:
Removed, added a tip to the partial to point to the RBAC for Database Access page to modify existing users.

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
````diff
@@ -0,0 +1,248 @@
+   user "teleport-user" has been updated
+   ```
+
+## Step 3/5. Create Role Mapping
````
```suggestion
## Step 3/5. Create a role mapping
```
Since we use sentence case in headings

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+
+Define a role mapping in Elasticsearch to assign your Teleport user or role to an Elasticsearch role.
+The example below maps the Teleport user `alice` to the `superuser` role in Elasticsearch, which
+should **not** be done in production:
```
I know it's implied, but I think it would help here to add a note about the kinds of roles a Teleport user should be allowed to have in Elasticsearch.

alexfornuto replied:
CC @Tener to suggest what roles would be appropriate.

Tener replied:
This is not a simple thing; the relevant docs are pretty extensive: https://www.elastic.co/guide/en/elasticsearch/reference/current/authorization.html

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+We will show you how to use `tctl auth sign` below.
+
+To securely access your Elasticsearch database, sign the certificate for the hostname Teleport will
+connect to:
```
```suggestion
connect to.
```

ptgott commented on docs/pages/database-access/guides/elastic.mdx:
````diff
@@ -0,0 +1,248 @@
+```
+
+<Admonition type="tip">
+For more information on configuring security settings in Elasticsearch, see:
````
I think this text is a better fit for the "Next steps" section. Per the [style guide](https://goteleport.com/docs/contributing/documentation/style-guide/#expectations-1), we're trying to avoid links in how-to guides so the user can get through the steps as quickly as possible

alexfornuto replied:
Added to the Next Steps list.

Tener commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+  --auth-server=teleport.example.com:3080 \
+  --name=myelastic \
+  --protocol=redis \
+  --uri={ELASTICSEARCH_HOST}:9200 \
```
It may be unclear that `{ELASTICSEARCH_HOST}` is meant to be replaced with an actual string like `localhost` or `elasticsearch.example.com`. I wonder if using an example hostname would be better, i.e. `elasticsearch.example.com` or similar.

alexfornuto replied:
replaced instances of {ELASTICSEARCH_HOST}` with `elasticsearch.example.com`. While I was at it, replaced similar password stand-in with `your_elasticsearch_password`

Tener commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,248 @@
+
+Alternately, you can edit an existing user with database access.
+
+1. As a Teleport user with the `editor` role, import a user's configuration to a local `yaml` file:
```
This section isn't specific to Elasticsearch and we don't have it in other guides. There are also other ways to achieve access given general flexibility of Teleport RBAC. It also won't work for SSO users. 

It is important to ensure that user can access the database, but perhaps we can delegate this to another article that can discuss the problem in depth? This section could then be replaced with instructions for quick access check (`tsh db ls`) and a link to the aforementioned article.

alexfornuto replied:
It looks like we have such a page, "Database Access Access Controls". Since a link to that page is at the bottom of this one, and with @ptgott's comment above, I think the best action is to remove this section, and possibly include a note in the `create-user.mdx` partial referencing it.

Tener commented on docs/pages/database-access/guides/elastic.mdx:
````diff
@@ -0,0 +1,248 @@
+```code
+$ curl -u elastic:{ELASTICSEARCH_USER_PASSWORD} -X POST "https://{ELASTICSEARCH_HOST}:9200/_security/role_mapping/mapping1?pretty" -H 'Content-Type: application/json' -d'
+{
+  "roles": [ "superuser"],
````
This is fine for testing in dev instance, but I'm wary of suggesting `superuser` for production usage. There are also other ways of setting role mapping in Elasticsearch, maybe we should link to Elasticsearch docs instead?

alexfornuto replied:
Please see the comment above.

Tener commented on docs/pages/database-access/guides/elastic.mdx:
````diff
@@ -0,0 +1,248 @@
+To connect to a particular database instance:
+
+```code
+$ tsh tb connect myelastic --db-user=elastic
````
```suggestion
$ tsh db connect myelastic --db-user=alice
```

alexfornuto replied:
I fixed the `tb` typo, but the question about `--db-user` remains.

Tener replied:
@alexfornuto Apologies, I missed your comment. `elastic` is a special built-in user and unlikely to be used by regular end users. We also added a mapping granting `alice` `superuser` role. The username which Elasticsearch sees - here `alice` - is whatever is provided to `--db-user`. The Teleport user must be allowed to use that `--db-user` and we ensure that with appropriate traits/roles combination.

alexfornuto commented on docs/pages/database-access/introduction.mdx:
```diff
@@ -16,13 +16,14 @@ Some of the things you can do with Database Access:
-- PostgreSQL
-- MySQL
+- CockroachDB
+- Elastic
```
```suggestion
- Elasticsearcg
```

alexfornuto commented on docs/pages/includes/database-access/guides.mdx:
```diff
@@ -1,20 +1,21 @@
-- [Self-hosted MongoDB](../../database-access/guides/mongodb-self-hosted.mdx): Connect self-hosted MongoDB database.
 - [Self-hosted CockroachDB](../../database-access/guides/cockroachdb-self-hosted.mdx): Connect self-hosted CockroachDB database.
-- [Self-hosted Redis](../../database-access/guides/redis.mdx): Connect self-hosted Redis.
+- [Self-hosted Elastic](../../database-access/guides/elastic.mdx)
```
```suggestion
- [Self-hosted Elasticsearch](../../database-access/guides/elastic.mdx)
```

Commit:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>
Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

Commit:
additional edits from review

Commit:
add spaces to next steps include

This is so that additional ul items added to a guide using this partial will have consistent spacing

alexfornuto commented on docs/pages/includes/database-access/guides-next-steps.mdx:
```diff
@@ -1,4 +1,7 @@
 - Learn how to [restrict access](../../database-access/rbac.mdx) to certain users and databases.
+
 - View the [High Availability (HA)](../../database-access/guides/ha.mdx) guide.
+
 - Take a look at the YAML configuration [reference](../../database-access/reference/configuration.mdx).
+
 - See the full CLI [reference](../../database-access/reference/cli.mdx).
```
Per the [commit message](https://github.com/gravitational/teleport/pull/17393/commits/938b5cf8fc625024d6271d431b1b1cef1f7e0647), I did this so additional ul items added will have consistent spacing.

Commit:
remove instructions and add tip

Resolves #r995144908 and #r995566035

Commit:
fix Database Access config and add scopes

Commit:
Move note into relevant tab

Commit:
adjust example user mapping

Commit:
incorporate more feedback

Tener commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,293 @@
+
+Once mutual TLS has been enabled, you will no longer be able to connect to the cluster without
+providing a valid client certificate. You can set `xpack.security.http.ssl.client_authentication`
+to `optional` to allow connections from clients that do not present a certificate.
```
May be worth adding that without client certificate the user must authenticate using a username and password (unless a guest account is enabled, in which case they will have guest privileges).

Tener commented on docs/pages/database-access/guides/elastic.mdx:
````diff
@@ -0,0 +1,293 @@
+To connect to a particular database instance:
+
+```code
+$ tsh db connect myelastic --db-user=elastic
````
The Elasticsearch role mapping above is for `alice`, but we are using `elastic` here.

Tener commented on docs/pages/database-access/guides/elastic.mdx:
````diff
@@ -0,0 +1,293 @@
+$ tsh db logout
+```
+
+## Next steps
````
I think a lot of people will use graphical clients, it might be worth showing how to use e.g. https://elasticvue.com/ combined with `tsh proxy db --tunnel`.

alexfornuto replied:
Added a new section above.

Commit:
incorporate feedback from @tener

alexfornuto commented on docs/pages/database-access/guides/elastic.mdx:
```diff
@@ -0,0 +1,293 @@
+
+(!docs/pages/includes/database-access/rotation-note.mdx!)
+
+Use the generated secrets to enable mutual TLS in your `elasticsearch.yml` configuration
```
I removed the preceding phrase.

Tener commented on docs/pages/database-access/guides/elastic.mdx:
````diff
@@ -0,0 +1,315 @@
+
+```code
+$ tsh proxy db myelastic --db-user=alice --tunnel
+Started authenticated tunnel for the  database "myelastic" in cluster "teleport.example.com" on 127.0.0.1:53657.
````
```suggestion
Started authenticated tunnel for the Elasticsearch database "myelastic" in cluster "teleport.example.com" on 127.0.0.1:53657.
```

`tsh` in 11.0.0-rc1 fixes the missing "Elasticsearch", not sure why it was missing in the first place.

alexfornuto replied:
@Tener can you link the PR? I'd like to see if it was or will be backported to 10.

Tener replied:
I don't have the PR, I just noticed that 11.0.0-alpha1 was missing "Elasticsearch" here while the latest version did not. V10 should behave the same.

Commit:
Update docs/pages/database-access/guides/elastic.mdx

Co-authored-by: Krzysztof Skrzętnicki <krzysztof.skrzetnicki@goteleport.com>

Commit:
Merge branch 'master' into afornuto/16963-elastic

Commit:
bypass linter rule

Commit:
Merge branch 'master' into afornuto/16963-elastic

Commit:
Merge branch 'master' into afornuto/16963-elastic
